lib/spack/spack/build_systems/perl.py

```diff
@@ -28,6 +30,41 @@ class PerlPackage(spack.package_base.PackageBase):
 
     extends("perl", when="build_system=perl")
 
+    @property
+    def use_modules(self) -> Iterable[str]:
+        """Names of the package's perl modules."""
+        module_files = find(self.prefix.lib, ["*.pm"], recursive=True)
+        # Drop the extension and library path
+        modules = [os.path.splitext(m)[0].replace(self.prefix.lib, "") for m in module_files]
+        # Drop the perl subdirectory as well
+        return ["::".join(m.split(os.sep)[2:]) for m in modules]
```

Comment:
This will not work for modules that are platform specific. As an example, a listing of the lib folder of YAML::Syck:

```
ls -R:

perl5

./perl5:
x86_64-linux-thread-multi

./perl5/x86_64-linux-thread-multi:
JSON  YAML  auto  perllocal.pod

./perl5/x86_64-linux-thread-multi/JSON:
Syck.pm

./perl5/x86_64-linux-thread-multi/YAML:
Dumper  Loader  Syck.pm
...

```

This will find './perl5/x86_64-linux-thread-multi/YAML/Syck.pm' and try to use it as:
```
use x86_64-linux-thread-multi::YAML::Syck;
```

which gives the error:
```
Bareword "thread" not allowed while "strict subs" in use at -e line 1.
```

It is not that easy to filter out that folder.
Some example names:
```
x86_64-linux-thread-multi
darwin-thread-multi-2level
darwin-64int-2level
i386-linux
```

Ideas on how to filter that folder:
Things that don't work:
- Arch folder starts with lower case. But Perl modules can be lower case too, e.g. autodie::variables

Things that may work:
- If the Perl arch name matches the Spack arch name, we could look for that. Not sure if it always does though.
- I think Perl distribution names can not have a dash, only an underscore. I also think that the arch folder must have at least one dash, but not really sure about that.

Things that should work but are more work to implement:
- For every Perl module we find, open the file and (try to) extract the 'package Foo'; line.
Then use that as the module name.
Of course, there may be multiple of those and various complications. But the simplistic approach should work for > 95% of the cases.

Reply:
I wasn't aware of that case.  Thanks!

Reply:
FYI. I am unable to install that particular package:

```
==> perl-yaml-syck: Executing phase: 'configure'
==> Error: ProcessError: Command exited with status 1:
    '/usr/bin/perl' 'Makefile.PL' 'INSTALL_BASE=$SPACK_ROOT/opt/spack/darwin-ventura-skylake/apple-clang-15.0.0/perl-yaml-syck-1.34-tes6q7amjezs75ydf5pr4rnthekjspdb'

1 error found in build log:
  >> 3    HASCOMPILER03EO/TESTEl3e.c:2:10: fatal error: 'EXTERN.h' file not fou
          nd
     4    #include "EXTERN.h"
     5             ^~~~~~~~~~
     6    1 error generated.
     7    Couldn't execute /Users/dahlgren1/releases/spack/lib/spack/env/clang/
          clang  -g -pipe -DPERL_USE_SAFE_PUTENV -Os "-I/System/Library/Perl/5.
          30/darwin-thread-multi-2level/CORE"   -c HASCOMPILER03EO/TESTEl3e.c -
          o HASCOMPILER03EO/TESTEl3e.o: Inappropriate ioctl for device at Makef
          ile.PL line 8.
     8    Sorry! YAML::Syck requires a compiler in order to be built.
```

Reply:
However, I was able to test the fix with `perl-threads` and `perl-scalar-list-utils`.